### ISSUE TITLE
/proc/self/fd/{0,1,2} are not valid (with --tty)

### ISSUE THREAD

**Issue by pwaller** [moby/moby#11462]
This is probably the root cause of #6880 and #8755.

Reproduced on 1.4.1 and 1.5.0.

Steps to reproduce:

```
$ docker run -ti --rm ubuntu ls -l /proc/self/fd
lrwx------ 1 root root 64 Mar 18 11:32 0 -> /25
lrwx------ 1 root root 64 Mar 18 11:32 1 -> /25
lrwx------ 1 root root 64 Mar 18 11:32 2 -> /25
lr-x------ 1 root root 64 Mar 18 11:32 3 -> /proc/1/fd
```

Expected: /fd/{0,1,2} symlinks should point to the same thing as `docker run -ti --rm ubuntu tty`.

Got: they point to some random point on the root of the filesystem which doesn't exist.

Effect: This breaks the command `tty` (and therefore tmux and various other programs) from a `docker exec` into the container. The reason is that `readlink(/proc/self/fd/0)` returns a path that doesn't exist. It also breaks any program which uses `/dev/std{in,out,err}`, because these are symlinks to items in the `/proc/self/fd/` directory.

**Comment by phemmer** [comment]
It looks like docker is using the host's `/dev/pts` mount for containers. All TTYs that docker creates aren't showing up in `/dev/pts`, only TTYs created once inside the container (tmux/screen/etc).

Note that these "symlinks" aren't real symlinks. You can still use them (`echo foo > /proc/self/fd/1`) as they're magic files handled by the kernel. They just link to a TTY outside the container, so they can't be properly handled by readlink(2).

```
# docker run -ti fedora:20 sh -c 'echo foo > /proc/self/fd/2'
foo

# docker run -ti fedora:20 sh -c 'echo foo > /dev/stderr'    
foo
```

So basically unless something is trying to perform a readlink(2) call, this shouldn't matter.

**Comment by pwaller** [comment]
Yes, `tty` tries to `readlink`. I also suspect that `tmux` does as well.

**Comment by cyphar** [comment]
For me, `/proc/self/fd/{0,1,2}` point to `/dev/pts/5` which doesn't exist (only `/dev/pts/ptmx` exists). Should we be mounting `pts` inside `/dev/pts`? Or just symlinking to `/dev/std{in,out,err}`?

**Comment by phemmer** [comment]
When docker is creating the container, it's launching the main process with a TTY that was created using the host's devpts mount. Thus things like `/proc/$pid/fd/0` are pointing to where the TTY is on the host's devpts mount.
Docker isn't doing any magic here, it's simply mounting a new devpts inside the container, so its contents don't match the host's. Also, docker also doesn't control where `/proc/$pid/fd/$fd` points to. That's controlled by the kernel.
`/dev/stdin` (and stdout/stderr) however are just vanilla symlinks that point to `/proc/self/fd/0` (and 1/2).
`/proc/$pid/fd/$fd` are "magic"ish. When you open one of these files (talking about the system level `open()` call), the kernel is duplicating that process's file descriptor. It's not resolving the symlink and opening what it points to.

The only (yes, only, there is no other) way to solve this (meaning to have `/proc/$pid/fd/$fd` point to valid paths in `/dev/pts`) is for the process that docker launches inside the container, to have a TTY that was created using the container's devpts mount.

**Comment by cyphar** [comment]
@phemmer Okay, so we need to create the TTY once we've switched to the new mount namespace (and mounted `devpts`)? Also, yeah the kernel does some pretty magical stuff in their userland `/proc` and `/sys` APIs...

**Comment by LK4D4** [comment]
@cyphar but then you need to send one side of pts to parent. I think that's why we doing it outside.

**Comment by cyphar** [comment]
@LK4D4 We don't need a PTMX in the parent, we just need to get the io to work, right? Why don't we just make pipes in the parent, then relay the PTMX output to the child ends?

**Comment by LK4D4** [comment]
@cyphar dunno, how we can connect pipe with ptmx in child?

**Comment by cyphar** [comment]
It's just an fd, we can just do a `go io.Copy(...)`. I'm just worried that there could be some magical things that break inexplicably with pipes (I mean, I'd never think that `ttyname` would cause so many false positive issues).

**Comment by LK4D4** [comment]
@cyphar I don't think that having any goroutine running in container is good idea. At least not when other solution is possible :)

**Comment by cyphar** [comment]
Yeah, you're right. Last thing we want is some mystical Go process running in a container. I really wish there was a way to fuse pipes in the kernel, that'd be really nice.

**Comment by pwaller** [comment]
Ouch. I've read docker's and the kernel sources and now I understand the problem reasonably thoroughly.

Just thinking around the problem, is it possible to mount the new `devpts -o newinstance` before we change into the mount namespace? If so, can we mount the "devpts newinstance" anywhere on the host temporarily, make the console that we'll use to talk to the containerised process whilst still on the host mount NS, then create the container NS and bind mount the new devpts into the container? Or is there a reason that won't work?

**Comment by LK4D4** [comment]
@pwaller I thought it's exactly what we doing, no? :)

**Comment by pwaller** [comment]
@LK4D4, unless I have the wrong end of the stick, I don't think so!

I'm suggesting this order:
1) Mount *newinstance* devpts
2) Make console from that
3) Put new devpts into new mount namespace (this is the step I'm unsure if it's possible)

This way, the console is visible on the host, and exists in the container namespace as needed.

The current order, as I understand it:

1) [Make console using *host* devpts mount](https://github.com/docker/docker/blob/7b0666f568ac46f10b328c18b7090e563593f88e/vendor/src/github.com/docker/libcontainer/standard_init_linux.go#L27)
2) [setupRootfs then follows a bit later](https://github.com/docker/docker/blob/7b0666f568ac46f10b328c18b7090e563593f88e/vendor/src/github.com/docker/libcontainer/standard_init_linux.go#L52)

(and as far as I can tell, the *newinstance* devpts is created inside there).

**Comment by pwaller** [comment]
There is another problem to think about, which is, how does one make `docker exec -t` work correctly? I guess you need the `newinstance` devpts mount to exist both on the host and inside the container.

**Comment by cyphar** [comment]
@pwaller For 3, we can just bind-mount it. But really, I'm not sure how that is any different from just making it in the host's devpts mount and then bind-mounting it. What you *actually* want is for us to make the console in the child's mount namespace -- the issue with that being that we can't access the console from outside in that case.

**Comment by pwaller** [comment]
From my understanding, its not the *mount* namespace that is the problem, it is the *devpts* namespace, which is created via the devpts -o newinstance mount. Since that mount can be bind mounted in two places, the problem is solvable, no? I suggest reading Documentation/.../devpts.txt in the kernel sources if you haven't. From memory so might require some looking. A key fact is that the new devpts mount gets its own /ptmx which makes the console in that namespace.  

On 3 July 2015 09:05:40 BST, Aleksa Sarai <notifications@github.com> wrote:
>@pwaller For 3, we can just bind-mount it. But really, I'm not sure how
>that is any different from just making it in the host's devpts mount
>and then bind-mounting it. What you *actually* want is for us to make
>the console in the child's mount namespace -- the issue with that being
>that we can't access the console from outside in that case.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/docker/docker/issues/11462#issuecomment-118269520

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by pwaller** [comment]
I had another look. It seems hard to fix, since it means that the `devpts` mount has to be made before the console is created from the `ptmx`. That means that the `devpts` mount (to some arbitrary path) probably needs to be done docker-daemon side (and then bind mounted into the right place by `standard_init_linux.go`).

So right now we have:

1. [`setupPipes()`](https://github.com/docker/docker/blob/5297a25b4eabb0bd6b76a09ea8f63d3e18329243/daemon/execdriver/native/driver.go#L127) which calls:
1. [`newConsole()`](https://github.com/docker/docker/blob/5297a25b4eabb0bd6b76a09ea8f63d3e18329243/vendor/src/github.com/opencontainers/runc/libcontainer/console_linux.go#L16), which is the non-containerized `/dev/ptmx` (which thus creates the console).
1. Later, the `devpts` mount happens inside [`setupRootfs()` in `standard_init_linux`](https://github.com/docker/docker/blob/5297a25b4eabb0bd6b76a09ea8f63d3e18329243/vendor/src/github.com/opencontainers/runc/libcontainer/standard_init_linux.go#L52) (which creates the new console namespace).

What we need is, when starting a container:

1. Mount new `dev/pts` somewhere (creating the new console namespace)
1. Use that new `dev/pts/ptmx` to [make a new console](https://github.com/docker/docker/blob/5297a25b4eabb0bd6b76a09ea8f63d3e18329243/vendor/src/github.com/opencontainers/runc/libcontainer/console_linux.go#L16)
1. Pass the path of the new `devpts` into `standard_init_linux`, which bind mounts it to `/dev/pts`

Then, `docker exec` would need to reuse the existing container's `dev/pts`, so the new `devpts` mount (the `devpts/ptmx`, specifically) would need to be available on the host.

Arranging this looks like it will require some fairly deep surgery.

**Comment by aidanhs** [comment]
If I'm understanding correctly, the downside of the above approach is it requires extra bookkeeping for runc (i.e. the need to keep a mountpoint around) which currently just keeps a config.json on a tmpfs.

**Comment by pwaller** [comment]
AFAIU, The `/dev/pts` mount doesn't need to survive the container being stopped. (My interpretation of your interpretation of things). It's just that the mount needs to be made inside the first container which is started and it must be visible on the host (so that it can be subsequently bind mounted into containers for `docker exec`).

**Comment by aidanhs** [comment]
Well the mount point is created on the host and bind mounted into the container - my concern is that it won't be automatically cleaned up on container death, there will have to be some teardown.

**Comment by crosbymichael** [comment]
I think I figured out the root cause of this issue.  A fix for runc can be found here https://github.com/opencontainers/runc/pull/279 and i'll look into how to port this to docker containers and things without user namespaces.

**Comment by crosbymichael** [comment]
Yup, I found a fix for this and will open a PR soon and link to it.

**Comment by cpuguy83** [comment]
:cookie:

**Comment by crosbymichael** [comment]
Fix is here in runc/libcontainer, i'll update docker after it's merged.  https://github.com/opencontainers/runc/pull/280

**Comment by cyphar** [comment]
@crosbymichael Hang on, how does this fix the whole "we can't run `su` in a container because it thinks that `/proc/self/fd/{0,1,2}` are invalid symlinks"?

**Comment by crosbymichael** [comment]
Let me verify the `su` issue

**Comment by pwaller** [comment]
I've amended the title of this issue to clarify that this bug only exists with `-t`, and that it may be a different problem than is fixed by @crosbymichael above.

**Comment by cyphar** [comment]
@pwaller I'd recommend also including that this only happens with `docker exec`.

**Comment by pwaller** [comment]
@cyphar that is not true - look at the single step required to reproduce.

**Comment by cyphar** [comment]
@pwaller Oh right. Sorry, I got this confused with a different issue which is caused by the same problem.

**Comment by aidanhs** [comment]
A good read for anyone new to this issue: https://www.kernel.org/doc/Documentation/filesystems/devpts.txt

**Comment by cyphar** [comment]
opencontainers/runc#1018 fixes this issue, and has been a long time coming. There are still other TTY-related issues but this is the first step to getting it all working. Hopefully this patchset will be merged soon.

**Comment by cyphar** [comment]
opencotnainers/runc#1018 has been merged. This can be closed as soon as Docker vendors the latest commit of `runc`.

**Comment by crosbymichael** [comment]
This has been fixed by #33007.  Let me know if you run into any problems
